### PR TITLE
Automatic and general updater improvements

### LINKED ISSUE #47: Add permitted network type setting for updates
As per https://developer.android.com/reference/androidx/work/NetworkType

### LINKED ISSUE #155: Auto-update toggle
Accrescent should have a toggle that disables auto-updates. Instead, users should get a notification about a new version, at which point they can choose whether they want to update or not.

Unlike the blacklist mentioned here (https://github.com/accrescent/accrescent/issues/154), this would apply across the board, and not to specific apps.

The blacklist feature would be useful if someone is obtaining a specific app from a different source and doesn't want to get it from Accrescent, and the auto-update toggle will be for people who want to check what a new version offers before updating etc.

### LINKED ISSUE #157: Notification for when an app is automatically updated through Accrescent
When an app is updated through Accrescent automatically, you should get a notification that says something along the lines of "`app_name` has been updated successfully." or something similar.

Unattended updates are a godsend, but it would be nice to actually know when something is updated, so I'm not perplexed when the app changes in some way when I'm unaware that it has updated in the first place.

### PR DESCRIPTION
**Pull request by Bnyro**
* Preference to choose the updater network conditions (closes #47)
* Preference to not update apps automatically (closes #155)
* Show a notification when an updated has finished (closes #157)

### CODE REVIEW

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt [hunk 1]
```diff
@@ -65,11 +78,23 @@ fun SettingsScreen(modifier: Modifier = Modifier, viewModel: SettingsViewModel =
...
+            onSelectionChanged = {
+                coroutineScope.launch { viewModel.setUpdaterNetworkType(context, networkTypeValues[it]) }
+            }
+        )
         SettingGroupLabel(stringResource(R.string.about), Modifier.padding(top = 16.dp))
         Setting(
             label = stringResource(R.string.source_code),
             description = stringResource(R.string.source_code_desc),
-            modifier = Modifier.fillMaxWidth(),
+            modifier = Modifier.fillMaxWidth()
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: nit: Unnecessary formatting change

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt [hunk 2]
```diff
@@ -89,7 +114,7 @@ fun SettingGroupLabel(text: String, modifier: Modifier = Modifier) {
         text = text,
         modifier = modifier.semantics { heading() },
         color = MaterialTheme.colorScheme.primary,
-        style = MaterialTheme.typography.labelLarge,
+        style = MaterialTheme.typography.labelLarge
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: nit: Unnecessary formatting change

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt [hunk 3]
```diff
@@ -98,16 +123,16 @@ fun Setting(
     label: String,
     description: String,
     modifier: Modifier = Modifier,
-    content: @Composable () -> Unit,
+    fillMaxWidth: Boolean = false,
+    content: @Composable () -> Unit
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt:146]: nit: Unnecessary formatting change

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt [hunk 4]
```diff
@@ -98,16 +123,16 @@ fun Setting(
...
 ) {
     Row(
         modifier = modifier.semantics(mergeDescendants = true) {},
         horizontalArrangement = Arrangement.SpaceBetween,
-        verticalAlignment = Alignment.CenterVertically,
+        verticalAlignment = Alignment.CenterVertically
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: nit: Unnecessary formatting change

File: app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt [hunk 5]
```diff
@@ -2,20 +2,26 @@ package app.accrescent.client.workers
...
 import dagger.assisted.AssistedInject
+import java.time.Duration
 
 @HiltWorker
 class AutoUpdateWorker @AssistedInject constructor(
     @Assisted val context: Context,
     @Assisted val workerParams: WorkerParameters,
     private val repoDataRepository: RepoDataRepository,
-    private val packageManager: PackageManager,
+    private val packageManager: PackageManager
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt]: nit: Unnecessary formatting change

File: app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt [hunk 6]
```diff
@@ -34,4 +40,24 @@ class AutoUpdateWorker @AssistedInject constructor(
 
         return Result.success()
     }
+
+    companion object {
+        private const val UPDATER_WORK_NAME = "UPDATE_APPS"
+
+        fun enqueue(context: Context, networkType: NetworkType, forceUpdate: Boolean = true) {
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt]: Is there any reason we should have this `forceUpdate` parameter instead of simply always using `ExistingPeriodicWorkPolicy.UPDATE`? As is, `UPDATE` will always be used anyway.

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt [hunk 7]
```diff
@@ -65,11 +78,23 @@ fun SettingsScreen(modifier: Modifier = Modifier, viewModel: SettingsViewModel =
                 onCheckedChange = { coroutineScope.launch { viewModel.setDynamicColor(it) } },
             )
         }
+        val networkTypeNames = listOf(R.string.all, R.string.unmetered, R.string.metered, R.string.not_roaming)
+        val networkTypeValues = listOf(NetworkType.CONNECTED, NetworkType.UNMETERED, NetworkType.METERED, NetworkType.NOT_ROAMING)
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: I ask this out of ignorance, but is there any reason someone would want to restrict updates to only metered networks?

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt [hunk 8]
```diff
@@ -116,3 +141,60 @@ fun Setting(
...
+    Setting(
+        label = label,
+        description = description,
+        modifier = modifier
+            .clickable(
+                interactionSource = interactionSource,
+                indication = null
+            ) {
+            showDialog = true
+        },
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: nit: Formatting (align braces), i.e.,

```kotlin
) { showDialog = true },
```

or

```kotlin
) {
    showDialog = true
},
```

as either pass auto-formatting.

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt (hunk 7 above)
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: nit: Please wrap this at ~100 columns

File: app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt [hunk 9]
```diff
@@ -9,38 +12,65 @@ import androidx.work.NetworkType
...
         }
 
         return Result.success()
     }
 
+    private fun showUpdateNotification(index: Int, packageInfo: PackageInfo, repoData: AppRepoData) {
+        val notificationManager = NotificationManagerCompat.from(context)
+
+        val notification = NotificationCompat.Builder(context, UPDATE_AVAILABLE_CHANNEL)
+            .setContentTitle(packageInfo.packageName)
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt]: I think it would be more appropriate to display the application label here instead of the app ID since it's user-facing

File: app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt [hunk 10]
```diff
@@ -49,6 +51,28 @@ class AppInstallBroadcastReceiver : BroadcastReceiver() {
                     notificationManager.notify(sessionId, notification)
                 }
             }
+            PackageInstaller.STATUS_SUCCESS -> {
+                val notification =
+                    NotificationCompat.Builder(context, Accrescent.UPDATED_FINISHED_CHANNEL)
+                        .setSmallIcon(R.drawable.ic_baseline_file_download_done_24)
+                        .setContentTitle(context.getString(R.string.update_finished))
+                        .setContentText(context.getString(R.string.update_success, sessionInfo?.appPackageName))
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt]: From my testing, `sessionInfo` seems to sometimes (always?) resolve to `null`, resulting in the message "null has been updated successfully." To reproduce:

1. Install Connect You < 2.4 from [GitHub releases](https://github.com/Bnyro/ConnectYou/releases).
2. Update Connect You from within Accrescent.
3. A notification with the erroneous message appears.

**Review by lberrymage — CHANGES_REQUESTED** [review]
Hey, thanks for the PR, and sorry it look a while for me to get a look at this! Looks pretty good except for one bug, some questions I have, and minor formatting things.

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt (hunk 2 above)
Comment (Bnyro) [reply app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: Oh, looks like I've accidentally run ktlint on that file. Is there any formatter Accrescent uses so that I can just use that one next time? Or is it just the inbuilt from Android Studio?

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt (hunk 7 above)
Comment (Bnyro) [reply app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: I can't come up with a viable use case for that, however I don't see a reason to not add the possibility since there are no disadvantages from making it possible.

File: app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt (hunk 6 above)
Comment (Bnyro) [reply app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt]: Oh damn, that's my bad, I wanted it to default to false.
In the `MainActivity` `forceUpdate` should be false so that the update checker timeout doesn't get reset when the user opens the app, since otherwise it would always be reset to `Connected` as required network type, no matter the user's preference (so my plans).
Though I'm not really sure whether it's the intended behavior that the updater resets (so the four hours or something start again from the beginning) or not, so that depends on your preference :)

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt (hunk 7 above)
Comment (matchboxbananasynergy) [reply app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: Personally, I don't see a use case here. I think that at least initially, following the equivalent settings in GrapheneOS' "Apps" app would make sense. Those are:

- Any
- Not roaming
- Unmetered

File: app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt (hunk 10 above)
Comment (Bnyro) [reply app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt]: Hmm, unfortunately I don't have any experience with the `PackageManager` so I've no clue why that happens. I decided to now pass the app's name directly to the broadcast receiver, which now works properly. However, it still seems to always use the app's package Name and not it's label although we're explicitely grabbing that from the `PackageManager` :/

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt (hunk 2 above)
Comment (lberrymage) [reply app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: Just the inbuilt one from Android Studio: https://github.com/accrescent/accrescent/blob/master/CONTRIBUTING.md#code-style. Ideally that would be automated in CI, but I haven't looked into it yet.

File: app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt (hunk 7 above)
Comment (lberrymage) [reply app/src/main/java/app/accrescent/client/ui/SettingsScreen.kt]: I'd rather not have the metered option if there isn't a good reason for it. It can always be added later if a reason is found, but for now I think it would just be confusing.

File: app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt (hunk 6 above)
Comment (lberrymage) [reply app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt]: The timer doesn't get reset with `UPDATE`. From https://developer.android.com/reference/androidx/work/ExistingPeriodicWorkPolicy#UPDATE:

> It preserves enqueue time, e.g. if a work was run 3 hours ago and had 8 hours long period, after the update it would be still eligible for run in 5 hours, assuming that periodicity wasn't updated.

[`CANCEL_AND_REENQUEUE`](https://developer.android.com/reference/androidx/work/ExistingPeriodicWorkPolicy#CANCEL_AND_REENQUEUE) would do what you're describing (which I agree would be undesirable).

File: app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt (hunk 10 above)
Comment (lberrymage) [reply app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt]: > However, it still seems to always use the app's package Name and not it's label although we're explicitely grabbing that from the `PackageManager`

Can you clarify? I don't understand. It appears to work fine now from my testing.

However, I think it would be cleaner to pull the package label from the [database](https://github.com/accrescent/accrescent/blob/0534277584ad49a6dd9b5d9e34f3026570116535/app/src/main/java/app/accrescent/client/data/db/App.kt#L10) by injecting `RepoDataRepository` into `AppInstallStatusReceiver` rather than extracting it from the package info and passing it via extras.

File: app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt (hunk 6 above)
Comment (Bnyro) [reply app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt]: Oh okay, I should better read the documentation next time ;)

Comment (Bnyro) [reply app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt]: However, the boolean flag is needed now since the call to enqueue the updater in the `MainActivity` should only enqueue the worker if no worker has been enqueued yet!

File: app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt (hunk 10 above)
Comment (Bnyro) [reply app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt]: > Can you clarify? I don't understand. It appears to work fine now from my testing.
>
It always shows the package name (e.g. com.bnyro.contacts) instead of the actual app name (Connect You). 
> However, I think it would be cleaner to pull the package label from the [database](https://github.com/accrescent/accrescent/blob/0534277584ad49a6dd9b5d9e34f3026570116535/app/src/main/java/app/accrescent/client/data/db/App.kt#L10) by injecting `RepoDataRepository` into `AppInstallStatusReceiver` rather than extracting it from the package info and passing it via extras.
>
Good idea, that will also resolve my issue!

File: app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt [hunk 11]
```diff
@@ -49,10 +62,42 @@ class AppInstallBroadcastReceiver : BroadcastReceiver() {
...
+
+                notificationManager.notify(sessionId, notification)
+            }
+            PackageInstaller.STATUS_FAILURE -> {
+                if (!context.hasNotificationPrivileges()) return
+
+                val notification =
+                    NotificationCompat.Builder(context, Accrescent.UPDATED_FINISHED_CHANNEL)
+                        .setSmallIcon(R.drawable.ic_baseline_error_outline_24)
+                        .setContentTitle(context.getString(R.string.update_finished))
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/receivers/AppInstallBroadcastReceiver.kt]: I think you meant to use `R.string.update_failure` here.

File: app/src/main/java/app/accrescent/client/Accrescent.kt [hunk 12]
```diff
@@ -30,27 +25,27 @@ class Accrescent : Application(), Configuration.Provider {
...
+                    getString(R.string.user_action_required),
+                    NotificationManager.IMPORTANCE_DEFAULT
+                ),
+                NotificationChannel(
+                    UPDATE_AVAILABLE_CHANNEL,
+                    getString(R.string.update_available),
+                    NotificationManager.IMPORTANCE_DEFAULT
+                ),
+                NotificationChannel(
+                    UPDATED_FINISHED_CHANNEL,
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/Accrescent.kt]: nit: Typo: "UPDATED" -> "UPDATE"

File: app/src/main/java/app/accrescent/client/Accrescent.kt [hunk 13]
```diff
@@ -73,5 +68,7 @@ class Accrescent : Application(), Configuration.Provider {
     companion object {
         lateinit var appContext: Context
         const val USER_ACTION_REQUIRED_CHANNEL = "UserActionRequired"
+        const val UPDATE_AVAILABLE_CHANNEL = "UpdateAvailable"
+        const val UPDATED_FINISHED_CHANNEL = "UpdateFinished"
```
Comment (lberrymage) [inline app/src/main/java/app/accrescent/client/Accrescent.kt]: nit: Typo: "UPDATED_FINISHED_CHANNEL" -> "UPDATE_FINISHED_CHANNEL"

**Review by lberrymage — CHANGES_REQUESTED** [review]
Two more typos I missed the first time around, but otherwise looks good.

I'm not sure why GitHub CI is failing because I can't reproduce the lint issue locally, but it seems fine so I can figure that out after merging.

**Comment by lberrymage** [comment]
Awesome! If you could (last thing, I promise!) rebase the edits back into the original 3 commits that'd be great. Otherwise I can do that if you like.

**Comment by Bnyro** [comment]
Alright, I'll do that in a minute!

File: app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt (hunk 6 above)
Comment (lberrymage) [reply app/src/main/java/app/accrescent/client/workers/AutoUpdateWorker.kt]: The worker isn't enqueued in `MainActivity`, so I assume you mean to refer to the application class.

I'm not sure where you're seeing an issue or how the flag solves it? It's fine to re-enqueue an existing periodic worker; that's what `ExistingPeriodicWorkPolicy` is for.

Just noting this for future reference. I'd like to get this merged soon, so I can change that myself later.
